From eb3b7d9ed388c0057cfd877d4e4dd86dfb733070 Mon Sep 17 00:00:00 2001
From: "Verma, Aashish" <aashishx.verma@intel.com>
Date: Fri, 7 Feb 2020 15:33:54 +0800
Subject: [PATCH 2/2] net: stmmac: fix missing IFF_MULTICAST check in
 dwmac4_set_filter
References: git-fixes
Git-commit: 2ba31cd93784b61813226d259fd94a221ecd9d61
Patch-mainline: v5.6-rc1

Without checking for IFF_MULTICAST flag, it is wrong to assume multicast
filtering is always enabled. By checking against IFF_MULTICAST, now
the driver behaves correctly when the multicast support is toggled by below
command:-
  ip link set <devname> multicast off|on

Fixes: 477286b53f55 ("stmmac: add GMAC4 core support")
Signed-off-by: Verma, Aashish <aashishx.verma@intel.com>
Tested-by: Tan, Tee Min <tee.min.tan@intel.com>
Signed-off-by: Ong Boon Leong <boon.leong.ong@intel.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Signed-off-by: Denis Kirjanov <denis.kirjanov@suse.com>
---
 drivers/net/ethernet/stmicro/stmmac/dwmac4_core.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/stmicro/stmmac/dwmac4_core.c b/drivers/net/ethernet/stmicro/stmmac/dwmac4_core.c
index f4c60a98fc48..899c9c0b17ae 100644
--- a/drivers/net/ethernet/stmicro/stmmac/dwmac4_core.c
+++ b/drivers/net/ethernet/stmicro/stmmac/dwmac4_core.c
@@ -410,7 +410,7 @@ static void dwmac4_set_filter(struct mac_device_info *hw,
 		 */
 		writel(0xffffffff, ioaddr + GMAC_HASH_TAB_0_31);
 		writel(0xffffffff, ioaddr + GMAC_HASH_TAB_32_63);
-	} else if (!netdev_mc_empty(dev)) {
+	} else if (!netdev_mc_empty(dev) && (dev->flags & IFF_MULTICAST)) {
 		u32 mc_filter[2];
 		struct netdev_hw_addr *ha;
 
-- 
2.16.4

